x/ccv/consumer/keeper/relay.go

```diff
@@ -86,46 +85,36 @@ func (k Keeper) OnRecvVSCPacket(ctx sdk.Context, packet channeltypes.Packet, new
 // operations that resulted in validator updates included in that VSC have matured on
 // the consumer chain.
 func (k Keeper) QueueVSCMaturedPackets(ctx sdk.Context) {
-	store := ctx.KVStore(k.storeKey)
-	maturityIterator := sdk.KVStorePrefixIterator(store, []byte{types.PacketMaturityTimeBytePrefix})
-	defer maturityIterator.Close()
-
 	currentTime := uint64(ctx.BlockTime().UnixNano())
 
-	maturedVscIds := []uint64{}
-	for maturityIterator.Valid() {
-		vscId := types.IdFromPacketMaturityTimeKey(maturityIterator.Key())
-		if currentTime >= binary.BigEndian.Uint64(maturityIterator.Value()) {
-			// construct validator set change packet data
-			vscPacket := ccv.NewVSCMaturedPacketData(vscId)
-
-			// append VSCMatured packet to pending packets
-			// sending packets is attempted each EndBlock
-			// unsent packets remain in the queue until sent
-			k.AppendPendingPacket(ctx, types.ConsumerPacket{
-				Type: types.VscMaturedPacket,
-				Data: vscPacket.GetBytes(),
-			})
-
-			ctx.EventManager().EmitEvent(
-				sdk.NewEvent(
-					ccv.EventTypeVSCMatured,
-					sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
-					sdk.NewAttribute(ccv.AttributeChainID, ctx.ChainID()),
-					sdk.NewAttribute(ccv.AttributeConsumerHeight, strconv.Itoa(int(ctx.BlockHeight()))),
-					sdk.NewAttribute(ccv.AttributeValSetUpdateID, strconv.Itoa(int(vscId))),
-					sdk.NewAttribute(ccv.AttributeTimestamp, strconv.Itoa(int(currentTime))),
-				),
-			)
-
-			maturedVscIds = append(maturedVscIds, vscId)
-		} else {
-			break
+	for _, maturityTime := range k.GetElapsedPacketMaturityTimes(ctx) {
+		if currentTime < maturityTime.MaturityTime {
+			panic(fmt.Errorf("maturity time %d is greater than current time %d", maturityTime.MaturityTime, currentTime))
 		}
-		maturityIterator.Next()
-	}
+		// construct validator set change packet data
+		vscPacket := ccv.NewVSCMaturedPacketData(maturityTime.VscId)
+
+		// append VSCMatured packet to pending packets
+		// sending packets is attempted each EndBlock
+		// unsent packets remain in the queue until sent
+		k.AppendPendingPacket(ctx, types.ConsumerPacket{
+			Type: types.VscMaturedPacket,
+			Data: vscPacket.GetBytes(),
+		})
 
-	k.DeletePacketMaturityTimes(ctx, maturedVscIds...)
+		k.DeletePacketMaturityTimes(ctx, uint64(maturityTime.VscId))
```

Comment:
> Duplicate packets cannot be appended because we delete PacketMaturityTimes for after appending them?

No. We delete PacketMaturityTimes to cleanup the state. We already got all the elapsed maturity times when calling `k.GetElapsedPacketMaturityTimes`. Why would duplicate packets be appended? This would require `PacketMaturityTimes` to contain two entries for the same `vscID`, see `OnRecvVSCPacket`.

Reply:
Just asking, it was throwing me off while reading.

The packet queue persists until all data is sent. Sending may or may not happen during EndBlock.
I was just checking how `GetElapsedPacketMaturityTimes` plays into this and making sure that cannot have duplicate packets in the queue if the queue is not emptied for a while.

Reply:
`GetElapsedPacketMaturityTimes` is not dependent on the packet queue. It's a different subset of keys.

Reply:
That is true, but the issue I'm concerned about.

The packet queue is constructed by iterating over elapsed maturity times (`GetElapsedPacketMaturityTimes`). If you iterate over the same maturity times more than once (in subsequent blocks), the queue would contain duplicated entries (multiple entries for same VSCId).

To avoid having that happen, `DeletePacketMaturityTimes` is called inside the loop.

Since this is a non-issue, we can close this discussion.